lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -350,3 +353,201 @@ def _get(self, json_data: JSON, _) -> Response:
+                operation. We want to support idempotency: the Dataset should
+                not be altered unless the error_count is 0.
+        """
+        raise NotImplementedError()
+
+    def _post(self, json_data: JSON, _) -> Response:
+        """
+        Perform the requested POST operation, and handle any exceptions.
+
+        Args:
+            json_data: Type-normalized client JSON input
+            _: Original incoming Request object (not used)
+
+        Returns:
+            Response to return to client
+        """
+        klasname = self.__class__.__name__
+        dataset = Dataset.attach(
+            controller=json_data["controller"], name=json_data["name"]
+        )
+        owner = User.query(id=dataset.owner_id)
+        if not owner:
+            self.logger.error(
+                "Dataset owner ID {} cannot be found in Users", dataset.owner_id
+            )
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="Dataset owner not found")
+
+        # For publish, we check authorization against the ownership of the
+        # dataset that was selected rather than having an explicit "user"
+        # JSON parameter. This will raise UnauthorizedAccess on failure.
+        self._check_authorization(owner.username, json_data["access"])
+
+        # Build an Elasticsearch instance to manage the bulk update
+        elastic = elasticsearch.Elasticsearch([self.es_url])
+
+        try:
+            # Pass the assemble() generator to the bulk helper
+            results = streaming_bulk(elastic, self.generate_documents(json_data, dataset), raise_on_error=False)
+        except BulkIndexError as e:
+            self.logger.exception("{} update failed: {}", klasname, e)
+            abort(HTTPStatus.INTERNAL_SERVER_ERROR, message="INTERNAL ERROR")
+        except elasticsearch.TransportError as e:
+            # Elasticsearch returned a status >= 400
+            self.logger.exception(
+                "{} HTTP error {} from Elasticsearch request",
+                klasname,
+                e
+            )
+            abort(
+                HTTPStatus.BAD_GATEWAY,
+                message=f"Elasticsearch query failure {e.error} ({e.status_code})",
+            )
+        except elasticsearch.ConnectionError as e:
+            self.logger.exception(
+                "{}: connection refused during the Elasticsearch request: {}", klasname, str(e.info)
+            )
+            abort(
+                HTTPStatus.BAD_GATEWAY,
+                message="Network problem, could not reach Elasticsearch",
+            )
+        except elasticsearch.UnsupportedProductError as e:
+            self.logger.exception(
+                "{}: Python elasticsearch package is not compatible with Elasticsearch server: {}", klasname, str(e)
+            )
+            abort(
+                HTTPStatus.INTERNAL_SERVER_ERROR,
+                message="Elasticsearch server version is incompatible",
+            )
```

Comment:
I don't know how much of this might actually be necessary given the bulk helper and `raise_on_error=False`; I need to investigate that. For now I left it all in (copied with minor tweaks from `ElasticBase`).